src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,108 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double COEFFICIENT = 1.5;
+    private T[] values;
+    private int size;
+
+    public ArrayList() {
+        values = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private void resize() {
+        int newCapacity = (int) (values.length * COEFFICIENT);
+        T[] newArray = (T[]) new Object[newCapacity];
+        System.arraycopy(values, 0, newArray, 0, size);
+        values = newArray;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == values.length) {
+            resize();
+        }
+        values[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add " + value + " to " + index);
+        }
+        if (size == values.length) {
+            resize();
+        }
+        System.arraycopy(values, index, values, index + 1, size - index);
+        values[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null) {
+            throw new NullPointerException("List is null");
+        }
+        if (size + list.size() > values.length) {
+            int newCapacity = (int) (Math.max(size + list.size(), values.length * COEFFICIENT));
+            T[] newArray = (T[]) new Object[newCapacity];
+            System.arraycopy(values, 0, newArray, 0, size);
+            values = newArray;
+        }
+        for (int i = 0; i < list.size(); i++) {
+            T element = list.get(i);
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index is wrong");
+        }
+        return values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
```

Comment:
What do you think about extracting index check in separate method and reuse those method?